### PR TITLE
feat: MLS 1:1 conversations [WPB-2183]

### PR DESCRIPTION
**Pull request by PatrykBuniX**
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-2183" title="WPB-2183" target="_blank"><img alt="Story" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10815?size=medium" />WPB-2183</a>  [Web] [Clients] Select protocol for 1-1 conversation
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

When MLS is supported and both of the users in a connection support MLS protocol (`supported_protocols` field on user entity includes `"mls"`) we want to:
- establish 1:1 MLS conversation
- add the other user into the MLS group
- move all the events (messages) into the new conversation
- hide old proteus conversation

### CODE REVIEW

File: src/script/components/Conversation/ReadOnlyConversationMessage.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+          <Icon.Info />
+        </div>
+      </div>
+      <p className="readonly-message-header-label" data-uie-name="element-readonly-conversation">
+        <span
+          dangerouslySetInnerHTML={{
+            __html: mlsCompatibilityMessage,
+          }}
+        />
+        {state === CONVERSATION_READONLY_STATE.READONLY_ONE_TO_ONE_SELF_UNSUPPORTED_MLS && (
```
Comment (przemvs) [inline src/script/components/Conversation/ReadOnlyConversationMessage.tsx:56]: Move this to variable, You using it multiple times :)

File: src/script/components/Conversation/ReadOnlyConversationMessage.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  handleMLSUpdate: () => void;
+  displayName: string;
+}
+export const ReadOnlyConversationMessage: FC<ReadOnlyConversationMessageProps> = ({
+  state,
+  handleMLSUpdate,
+  displayName,
+}) => {
+  const mlsCompatibilityMessage =
+    state === CONVERSATION_READONLY_STATE.READONLY_ONE_TO_ONE_OTHER_UNSUPPORTED_MLS
```
Comment (przemvs) [inline src/script/components/Conversation/ReadOnlyConversationMessage.tsx:39]: Move this to variable, You using it multiple times :)

File: src/script/components/Conversation/ReadOnlyConversationMessage.tsx (hunk 1 above)
Comment (PatrykBuniX) [reply src/script/components/Conversation/ReadOnlyConversationMessage.tsx:56]: Those two checks you've commented on are actually different. Once is `READONLY_ONE_TO_ONE_SELF_UNSUPPORTED_MLS`, the other one is `READONLY_ONE_TO_ONE_OTHER_UNSUPPORTED_MLS` (`SELF` vs `OTHER`)

File: src/script/components/Conversation/ReadOnlyConversationMessage.tsx (hunk 2 above)
Comment (PatrykBuniX) [reply src/script/components/Conversation/ReadOnlyConversationMessage.tsx:39]: https://github.com/wireapp/wire-webapp/pull/15508#discussion_r1350414740

**Review by przemvs — APPROVED** [review]

File: src/i18n/en-US.json [hunk 3]
```diff
@@ -727,9 +727,14 @@
   "messageFailedToSendWillNotReceiveSingular": "won't get your message.",
   "messageFailedToSendWillReceivePlural": "will get your message later.",
   "messageFailedToSendWillReceiveSingular": "will get your message later.",
-  "mlsConversationRecovered": "You haven’t used this device for a while, or an issue has occurred. Some older messages may not appear here.",
+  "messageReactionDetails": "{{emojiCount}} reaction, react with {{emojiName}} emoji",
```
Comment (atomrc) [inline src/i18n/en-US.json]: hummm, this seems unrelated to the topic 🤔

File: src/script/conversation/ConversationMapper.ts [hunk 4]
```diff
@@ -178,6 +179,10 @@ export class ConversationMapper {
       conversationEntity.verification_state(verification_state);
     }
 
+    if (archived_state !== undefined) {
+      conversationEntity.archivedState(archived_state);
+    }
+
```
Comment (atomrc) [inline src/script/conversation/ConversationMapper.ts]: This seems somewhat unrelated 🤔

File: src/script/conversation/ConversationService.ts [hunk 5]
```diff
@@ -430,4 +430,17 @@ export class ConversationService {
...
+   * Will check if mls group is established locally.
+   * Group is established after the first commit was sent in the group and epoch number is at least 1.
+   * @param groupId groupId of the conversation
+   */
+  async isMLSGroupEstablishedLocally(groupId: string): Promise<boolean> {
+    return this.coreConversationService.isMLSGroupEstablishedLocally(groupId);
+  }
+
+  async getMLS1to1Conversation(userId: QualifiedId) {
+    return this.apiClient.api.conversation.getMLS1to1Conversation(userId);
```
Comment (atomrc) [inline src/script/conversation/ConversationService.ts]: don't we have a `core` method that would do that? As much as possible we should avoid using the `apiClient` directly in the webapp (in the long run the apiClient should be hidden to the webapp)

File: src/script/page/AppMain.tsx [hunk 6]
```diff
@@ -225,6 +225,7 @@ const AppMain: FC<AppMainProps> = ({
                 selfUser={selfUser}
                 isRightSidebarOpen={!!currentState}
                 openRightSidebar={toggleRightSidebar}
+                onRefresh={app.refresh}
```
Comment (atomrc) [inline src/script/page/AppMain.tsx]: suggestion: maybe the naming is not really explicit. I feel that the app will completely reload when something (I don't know what) is refreshed. 

maybe we could try to be more explicit `onUserSupportedProtocolUpdated` (I don't really know what is the trigger here, you would probably find a better name :D )

File: src/script/main/app.ts [hunk 7]
```diff
@@ -413,10 +414,6 @@ export class App {
         conversationRepository.initMLSConversationRecoveredListener();
       }
 
-      if (connections.length) {
-        await Promise.allSettled(conversationRepository.mapConnections(connections));
-      }
-
```
Comment (atomrc) [inline src/script/main/app.ts:419]: I guess this is done by the `init1To1Conversations` right? Maybe it would make sense to put it in another function or is this to tied to actual 1:1 conversations?

File: src/i18n/en-US.json (hunk 3 above)
Comment (PatrykBuniX) [reply src/i18n/en-US.json]: No idea what it's doing here, good catch, will remove it.

File: src/script/conversation/ConversationService.ts (hunk 5 above)
Comment (PatrykBuniX) [reply src/script/conversation/ConversationService.ts]: We do not have it yet but i will add it 👌. I like the approach you're suggesting 🙏

File: src/script/page/AppMain.tsx (hunk 6 above)
Comment (PatrykBuniX) [reply src/script/page/AppMain.tsx]: This one should not start with `"on"` actually, I suggest naming it `reloadApp`, it's called when user clicks an update button in a warning message we display when they do not support mls and the other user (the one we try to communicate with via 1:1 conversation) does support only mls (does not support proteus). After app is reloaded, we will re-evaluate 1:1 conversations again. There's a chance (but not guaranty), that after the reload self user's team has started supporting mls, so conversation can be finally established. 

So it's not a listener (`"onSomething"`), but rather an action (`"doSomething"`). Makes sense to you?

File: src/script/main/app.ts (hunk 7 above)
Comment (PatrykBuniX) [reply src/script/main/app.ts:419]: Correct, this logic was moved to `init1To1Conversations`, I don't quite get what you mean, you've commented on removed code 🤔

Comment (atomrc) [reply src/script/main/app.ts:419]: My question is: would it make sense to have a separate function that does the connection mapping (rather than doing it inside of the `init1to1conversations` method). In order to split reponsibilities. But maybe it doesn't make sense, I leave you judge of this :D

File: src/script/page/AppMain.tsx (hunk 6 above)
Comment (atomrc) [reply src/script/page/AppMain.tsx]: Sure, maybe this makes more sense 👍

File: src/script/main/app.ts (hunk 7 above)
Comment (PatrykBuniX) [reply src/script/main/app.ts:419]: Yeah, I'd leave it as it is now - we're initialising 1:1 conversation every time we map through connection (and find a 1:1 conversation for this connection), so i'd say it's tied to it. Right after that we need to map through team-owned 1:1 conversations (that don't have connections assigned) anyway, I prefer to have initApp method calling a single method that takes care of initialising all the 1:1 conversations.

**Review by atomrc — APPROVED** [review]
